src/Stratis.Bitcoin/Utilities/AsyncConsumerQueue.cs

```diff
@@ -0,0 +1,154 @@
+using System.Threading.Tasks;
+using NBitcoin;
+
+namespace Stratis.Bitcoin.Utilities
+{
+    /// <summary>
+    /// Represents a queue of items that are processed when a certain condition is met or when a timer runs out.
+    /// </summary>
+    public interface IAsyncConsumerQueue<T> : IDisposable
+    {
+        /// <summary>Adds object to the end of the queue.</summary>
+        void Enqueue(T item);
+    }
+
+    /// <inheritdoc />
+    public class AsyncConsumerQueue<T> : IAsyncConsumerQueue<T>
+    {
+        private readonly ConcurrentQueue<T> queue;
+
+        private readonly Action<ConcurrentQueue<T>> processingAction;
+        private readonly Func<T, bool> immediateProcessingTrigger;
+        private readonly TimeSpan? timer;
+        private readonly bool restartTimerIfNotAllItemsConsumed;
+
+        private readonly AsyncManualResetEvent trigger;
+
+        /// <summary>Task that waits for <see cref="trigger"/> to be triggered.</summary>
+        private Task triggerAwaiterTask;
+
+        /// <summary>Task that waits until specified delay runs out.</summary>
+        private Task timerAwaiterTask;
+
+        /// <summary>Task that runs execution flow.</summary>
+        private Task executor;
+
+        /// <summary>Cancellation token source.</summary>
+        private CancellationTokenSource cancellation;
+
+        private object lockObject;
+
+        /// <summary>Queue items count.</summary>
+        public int Count => this.queue.Count;
+
+        /// <summary>
+        /// Creates new instance of <see cref="AsyncConsumerQueue{T}<T>"/>.
+        /// </summary>
+        /// <param name="processingAction">Action that consumes items from the queue and removes them when they are consumed.</param>
+        /// <param name="immediateProcessingTrigger">Condition that has to be met in order to start consumption action immediately.</param>
+        /// <param name="timer">Maximum amount of time that can pass since adding item to the queue before it gets consumed.</param>
+        /// <param name="restartTimerIfNotAllItemsConsumed">If set to <c>true</c> timer will be automatically restarted if not all items were consumed.</param>
+        public AsyncConsumerQueue(
+            Action<ConcurrentQueue<T>> processingAction,
+            Func<T, bool> immediateProcessingTrigger = null,
+            TimeSpan? timer = null,
+            bool restartTimerIfNotAllItemsConsumed = true)
+        {
+            this.queue = new ConcurrentQueue<T>();
+            this.processingAction = processingAction;
+            this.immediateProcessingTrigger = immediateProcessingTrigger;
+            this.timer = timer;
+            this.restartTimerIfNotAllItemsConsumed = restartTimerIfNotAllItemsConsumed;
+
+            this.trigger = new AsyncManualResetEvent(false);
+            this.cancellation = new CancellationTokenSource();
+            this.lockObject = new object();
+        }
+
+        /// <inheritdoc />
+        public void Enqueue(T item)
```

Comment:
i don't think this implementation of enqueue is good

why would you run task on the background for this, it does not make sense to me

who don't you just enqueue the item and set the event on which the consumer task is waiting
that's literally two lines of code (+logs)

Reply:
>why would you run task on the background for this, it does not make sense to me

because in case 'immediateProcessingTrigger' condition is heavy we don't wan't to lock current thread checking it.